File: python/hashtable.py
import math
class HashTable:
    def __init__(self,capacity=1,a=1,scaling_factor=1,shift=0):
        self.capacity = capacity
        self.a = a
        self.scaling_factor = scaling_factor
        self.shift = shift
        self.S = []
        for i in range(self.capacity):
            self.S.append([])
        self.size = 0
    
    def size(self):
        return self.size
    
    def is_empty(self):
        return self.size > 0
    
    def hash_code(self,s):
        code = 7
        for i in range(len(s)):
            code = self.a*code  + ord(s[i])
        return code
    
    def hash(self,s):
        return int(math.fabs(self.scaling_factor*self.hash_code(s) + self.shift) % self.capacity)
    
    def put(self,s):
        i = self.hash(s)
        bucket = self.S[i]
        if (s not in bucket):
            bucket.append(s)
            self.size+=1
    
    def get(self,s):
        i = hash(s)
        bucket = self.S[i]
        return s in bucket

File: java/hashtables/HashTable.java
public class Hashtable {
    private int capacity;
    private Object[] S;
    private int a;
    private int scalingFactor;
    private int shift;
    private int size;

    public HashTable(int capacity,int a,int scalingFactor,int shift) {
        this.capacity = capacity;
        this.a = a;
        this.scalingFactor = scalingFactor;
        this.shift = shift;
        s = new Object[capacity];
        for (int i=0;i<capacity;i++) S[i] = new ArrayList<String>();
        size = 0;
    }

    public HashTable() {
        capacity = 1;
        a = 1;
        scalingFactor = 1;
        shift = 0;
        S = new Object[capacity];
        for (int i=0;i<capacity;i++) S[i] = new ArrayList<String>();
        size =0;
    }

    public int size() { return size; }

    public boolean isEmpty() { return size > 0; }

    private long hashCode(String s) {
        long h =7;
        for (int i=0;i<s.length();i++) h=a*h + s.charAt(i);
        return h;
    }

    private int hash(String s) {
        return (int) (Math.abs(scalingFactor*hashCode(s)+shift) % capacity);
    }

    public void put(String s) {
        int i = hash(s);
        ArrayList<String> bucket = (ArrayList<String>)S[i];
        if(!bucket.contains(s)) { bucket.add(s);size++;}
    }

    public boolean get(String s) {
        int i = hash(s);
        ArrayList<String> bucket = (ArrayList<String>)S[i];
        return bucket.contains(s);
    }

    

}